[Feature] Bump pydantic to >=2.0.0

### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward change to existing dbt-semantic-interfaces functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

This package currently blocks users of dbt-core from upgrading Python projects to pydantic 2.0.0 which has the well-documented performance improvements.

### Describe alternatives you've considered

_No response_

### Who will this benefit?

_No response_

### Are you interested in contributing this feature?

_No response_

### Anything else?

_No response_

Hi,

I wanted to help out, so I had a look.

The change does not seem straight forward at all.

@QMalcolm I tried to follow the same approach as you and started from [your branch on my own fork](https://github.com/esciara/dbt-semantic-interfaces/tree/qmalcolm--pydantic-2-support). I did the following:
- I rebased with `main` to be up-to-date.
- I updated the dependency to `pydantic~=2.5` to start with the latest version.
- I ran the `bump-pydantic` tool as suggested by @elongl .
- I organised the changes in commits, mimicking the way you did it to separate the changes in stages.
- I committed the code that still needs work (as suggested by `bump-pydantic`) in a "WIP" commit.

A lot of tests are not passing. I digged into some of them.

One of the main issue is that the deprecation of `__get_validators__` has a big impact, especially you used it to instantiate object during deserialisation by not using a `dict` or another instance of the same class,  but by using base types such as `str` for instance ([here is an example with `PydanticWhereFilter`](https://github.com/esciara/dbt-semantic-interfaces/blob/192881b290df06f4efd02079863c97ec191f1558/dbt_semantic_interfaces/implementations/filters/where_filter.py#L46-L47)). I could not find a way to emulate that with `__get_pydantic_core_schema__` (supposedly `__get_validators__`'s replacement).

Another issue with is that for some reason, validation fails on `XxxAsPydantic` classes as it thinks that the instances that are passed do not have the same type as expected, even though they seems to do... Probably has to do with them being created dynamically. Here is the result for running : 

```
test_dataclass_serialization.py::test_nested_dataclass FAILED            [100%]
test_dataclass_serialization.py:93 (test_nested_dataclass)

[..]

>           return PydanticModel(**field_values)
E           pydantic_core._pydantic_core.ValidationError: 1 validation error for NestedDataclassAsPydantic
E           field1
E             Input should be a valid dictionary or instance of SimpleDataclassAsPydantic [type=model_type, input_value=SimpleDataclassAsPydantic(field0=1), input_type=SimpleDataclassAsPydantic]
E               For further information visit https://errors.pydantic.dev/2.5/v/model_type

../dbt_semantic_interfaces/dataclass_serialization.py:223: ValidationError
```

Any guidance for a clue?

--- a/.changes/unreleased/Dependencies-20231221-130418.yaml
+++ b/.changes/unreleased/Dependencies-20231221-130418.yaml
@@ -1,0 +1,6 @@
+kind: Dependencies
+body: Add support for pydantic 2 (and continue support for pydantic 1)
+time: 2023-12-21T13:04:18.792592-08:00
+custom:
+  Author: QMalcolm bernardcooke53 esciara
+  PR: "134"

--- a/.github/workflows/ci-pre-commit.yaml
+++ b/.github/workflows/ci-pre-commit.yaml
@@ -16,5 +16,8 @@ jobs:
     steps:
       - uses: actions/checkout@v3
       - uses: ./.github/actions/setup-python-env
+      # This step is necessary so long as we're allowing Pydantic 1 and Pydantic 2 via shimming
+      - name: Force Pydantic 1
+        run: hatch run dev-env:pip install "pydantic~=1.10"
       - name: Run Pre-commit Hooks
         run: hatch run dev-env:pre-commit run --show-diff-on-failure --color=always --all-files

--- a/.github/workflows/ci-pytest.yaml
+++ b/.github/workflows/ci-pytest.yaml
@@ -11,11 +11,12 @@ on:
 
 jobs:
   pytest:
-    name: Run Tests / Python ${{ matrix.python-version }}
+    name: Run Tests / Python ${{ matrix.python-version }} / Pydantic ~= ${{ matrix.pydantic-version }}
 
     strategy:
       matrix:
         python-version: ["3.8", "3.9", "3.10", "3.11"]
+        pydantic-version: ["1.10", "2.0"]
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
@@ -26,5 +27,7 @@ jobs:
       - name: Install Hatch
         shell: bash
         run: pip3 install hatch
+      - name: Set pydantic Version ~= ${{ matrix.pydantic-version }}
+        run: hatch run dev-env:pip install "pydantic~=${{ matrix.pydantic-version }}"
       - name: Run Python Tests
         run: hatch run dev-env:pytest tests

--- a/dbt_semantic_interfaces/dataclass_serialization.py
+++ b/dbt_semantic_interfaces/dataclass_serialization.py
@@ -19,11 +19,10 @@ from typing import (
     get_type_hints,
 )
 
-import pydantic
-from pydantic import BaseModel
 from typing_extensions import TypeAlias
 
 from dbt_semantic_interfaces.pretty_print import pformat_big_objects
+from pydantic_shim import BaseModel, create_model
 
 logger = logging.getLogger(__name__)
 
@@ -347,7 +346,7 @@ class DataClassTypeToPydanticTypeConverter:  # noqa: D
         logger.debug(
             f"Creating Pydantic model {class_name} with fields:\n{pformat_big_objects(fields_for_pydantic_model)}"
         )
-        pydantic_model = pydantic.create_model(class_name, **fields_for_pydantic_model)  # type: ignore
+        pydantic_model = create_model(class_name, **fields_for_pydantic_model)  # type: ignore
         logger.debug(f"Finished creating Pydantic model {class_name}")
         logger.debug(f"Finished converting {dataclass_type.__name__} to a pydantic class")
         return pydantic_model

--- a/dbt_semantic_interfaces/implementations/base.py
+++ b/dbt_semantic_interfaces/implementations/base.py
@@ -5,13 +5,12 @@ import os
 from abc import ABC, abstractmethod
 from typing import Any, Callable, ClassVar, Generator, Generic, Type, TypeVar
 
-from pydantic import BaseModel, root_validator
-
 from dbt_semantic_interfaces.errors import ParsingException
 from dbt_semantic_interfaces.parsing.yaml_loader import (
     PARSING_CONTEXT_KEY,
     ParsingContext,
 )
+from pydantic_shim import BaseModel, root_validator
 
 # Type alias for the implicit "Any" type used as input and output for Pydantic's parsing API
 PydanticParseableValueType = Any  # type: ignore[misc]

--- a/dbt_semantic_interfaces/implementations/export.py
+++ b/dbt_semantic_interfaces/implementations/export.py
@@ -2,7 +2,6 @@ from __future__ import annotations
 
 from typing import Optional
 
-from pydantic import Field
 from typing_extensions import override
 
 from dbt_semantic_interfaces.implementations.base import HashableBaseModel
@@ -11,6 +10,7 @@ from dbt_semantic_interfaces.protocols.export import Export, ExportConfig
 from dbt_semantic_interfaces.type_enums.export_destination_type import (
     ExportDestinationType,
 )
+from pydantic_shim import Field
 
 
 class PydanticExportConfig(HashableBaseModel, ProtocolHint[ExportConfig]):

--- a/dbt_semantic_interfaces/implementations/metric.py
+++ b/dbt_semantic_interfaces/implementations/metric.py
@@ -2,8 +2,6 @@ from __future__ import annotations
 
 from typing import List, Optional, Sequence
 
-from pydantic import Field
-
 from dbt_semantic_interfaces.enum_extension import assert_values_exhausted
 from dbt_semantic_interfaces.errors import ParsingException
 from dbt_semantic_interfaces.implementations.base import (
@@ -22,6 +20,7 @@ from dbt_semantic_interfaces.type_enums import (
     MetricType,
     TimeGranularity,
 )
+from pydantic_shim import Field
 
 
 class PydanticMetricInputMeasure(PydanticCustomInputParser, HashableBaseModel):

--- a/dbt_semantic_interfaces/implementations/project_configuration.py
+++ b/dbt_semantic_interfaces/implementations/project_configuration.py
@@ -3,7 +3,6 @@ from __future__ import annotations
 from typing import List, Optional
 
 from importlib_metadata import version
-from pydantic import validator
 from typing_extensions import override
 
 from dbt_semantic_interfaces.implementations.base import (
@@ -20,6 +19,7 @@ from dbt_semantic_interfaces.implementations.time_spine_table_configuration impo
 )
 from dbt_semantic_interfaces.protocols import ProtocolHint
 from dbt_semantic_interfaces.protocols.project_configuration import ProjectConfiguration
+from pydantic_shim import validator
 
 
 class PydanticProjectConfiguration(HashableBaseModel, ModelWithMetadataParsing, ProtocolHint[ProjectConfiguration]):

--- a/dbt_semantic_interfaces/implementations/semantic_model.py
+++ b/dbt_semantic_interfaces/implementations/semantic_model.py
@@ -2,7 +2,6 @@ from __future__ import annotations
 
 from typing import Any, List, Optional, Sequence
 
-from pydantic import validator
 from typing_extensions import override
 
 from dbt_semantic_interfaces.implementations.base import (
@@ -25,6 +24,7 @@ from dbt_semantic_interfaces.references import (
     SemanticModelReference,
     TimeDimensionReference,
 )
+from pydantic_shim import validator
 
 
 class NodeRelation(HashableBaseModel):

--- a/dbt_semantic_interfaces/validations/validator_helpers.py
+++ b/dbt_semantic_interfaces/validations/validator_helpers.py
@@ -20,7 +20,6 @@ from typing import (
 )
 
 import click
-from pydantic import BaseModel, Extra
 
 from dbt_semantic_interfaces.implementations.base import FrozenBaseModel
 from dbt_semantic_interfaces.protocols import Metadata, SemanticManifestT, SemanticModel
@@ -30,6 +29,7 @@ from dbt_semantic_interfaces.references import (
     SemanticModelReference,
 )
 from dbt_semantic_interfaces.type_enums import DimensionType
+from pydantic_shim import BaseModel, Extra
 
 VALIDATE_SAFELY_ERROR_STR_TMPLT = ". Issue occurred in method `{method_name}` called with {arguments_str}"
 ValidationContextJSON = Dict[str, Union[str, int, None]]

--- a/pydantic_shim.py
+++ b/pydantic_shim.py
@@ -1,0 +1,26 @@
+from importlib.metadata import version
+
+pydantic_version = version("pydantic")
+# Pydantic uses semantic versioning, i.e. <major>.<minor>.<patch>, and we need to know the major
+pydantic_major = pydantic_version.split(".")[0]
+
+if pydantic_major == "1":
+    from pydantic import (  # type: ignore  # noqa
+        BaseModel,
+        Extra,
+        Field,
+        create_model,
+        root_validator,
+        validator,
+    )
+elif pydantic_major == "2":
+    from pydantic.v1 import (  # type: ignore  # noqa
+        BaseModel,
+        Extra,
+        Field,
+        create_model,
+        root_validator,
+        validator,
+    )
+else:
+    raise RuntimeError(f"Currently only pydantic 1 and 2 are supported, found pydantic {pydantic_version}")

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -20,7 +20,7 @@ classifiers = [
   "Programming Language :: Python :: Implementation :: PyPy",
 ]
 dependencies = [
-  "pydantic~=1.10",
+  "pydantic>=1.10,<3",
   "jsonschema~=4.0",
   "PyYAML~=6.0",
   "more-itertools>=8.0,<11.0",

